File: services/workers/src/workers/parser.py
"""Stage 2 — Parser worker: THE heavy one (PRD §6.3).

Memory discipline, all five controls required:
  one job per process (prefetch=1, concurrency=1)   — compose env
  process recycling (PARSER_RECYCLE_AFTER)          — compose + runner (clean exit per N jobs)
  soft RSS budget (SoftOOM)                         — parsing.memory
  hard container limit (mem_limit: 8g)              — compose backstop
  tmpfs cap (2g)                                    — compose

Retry ladder (attempts are not identical):
  1 as configured
  2 re-split this shard into 4 sub-shards
  3 single-page shards
  4 disable table structure, text-only
  final mark shard failed, emit event, continue the book
"""

from __future__ import annotations

import logging
import os
import shutil
import tempfile
import time
import uuid
from pathlib import Path

from core.config import Settings, get_settings
from core.db import repo
from core.errors import ErrorCode, PlatformError
from core.queue import contracts, streams
from core.storage import s3
from parsing.converter import (  # noqa: F401  (build_converter kept importable: tests patch it here)
    _CONVERTER_CACHE,
    build_converter,
    converter_cache_key,
    get_converter,
)
from parsing.memory import check_rss_budget
from parsing.ocr_gate import needs_ocr
from parsing.splitter import fixed_bounds as parsing_splitter_fixed_bounds


def _ladder_action(shard, s: Settings) -> str:
    """Attempt 2: quarter-split; attempt 3: single-page; attempt >=4:
    text-only convert. A shard too small to split skips straight to
    text-only. Return "split"|"text_only"."""
    span = shard.page_end - shard.page_start + 1
    if shard.attempts == 2 and span >= 4 * max(1, s.shard_pages // 4) // 2:  # worth quarter-splitting
        return "split"
    if shard.attempts == 3 and span >= 2:
        return "split"  # single-page sub-shards
    return "text_only"


def _split_and_requeue(session, redis, doc, shard, s: Settings) -> None:
    """Replace a failing shard with finer sub-shards (ladder attempts 2-3).
    The parent becomes SKIPPED; each sub-shard is a fresh ParseJob."""
    from core.events import write_event

    span = shard.page_end - shard.page_start + 1
    shard_pages = 1 if shard.attempts >= 3 else max(1, span // 4)
    bounds = [
        (shard.page_start + b.page_start - 1, shard.page_start + b.page_end - 1)
        for b in parsing_splitter_fixed_bounds(span, shard_pages, overlap=0)
        # overlap=0 is required: sub-shards must tile the parent disjointly
        # (fixed_bounds' default overlap=1 would emit overlapping bounds —
        # a span-20 attempt-2 shard would yield 5 overlapping bounds like
        # [1-5],[5-9],[9-13],... instead of 4 disjoint ones, double-counting
        # total_shards and re-parsing boundary pages).
    ]
    start_idx = repo.next_shard_idx(session, doc.id)
    repo.skip_shard(session, doc.id, shard.idx)
    repo.insert_shards(session, doc.id, bounds, start_idx=start_idx)
    doc.total_shards = (doc.total_shards or 0) + len(bounds)
    session.flush()
    for i, (ps, pe) in enumerate(bounds):
        streams.xadd_job(
            redis,
            streams.STREAM_PARSE,
            contracts.ParseJob(
                doc_id=doc.id,
                idx=start_idx + i,
                page_start=ps,
                page_end=pe,
                source_uri=doc.source_uri,
            ),
        )
    write_event(
        session,
        "warn",
        "parse",
        f"shard {shard.idx} re-split into {len(bounds)} sub-shards "
        f"(ladder attempt {shard.attempts})",
        doc_id=doc.id,
        shard_idx=shard.idx,
        code="SHARD_RESPLIT",
    )


def handle_parse(session, job: dict, redis, settings: Settings | None = None) -> None:
    s = settings or get_settings()
    doc_id = uuid.UUID(str(job["doc_id"]))
    idx = int(job["idx"])
    page_start = int(job["page_start"])
    page_end = int(job["page_end"])

    doc = repo.get_document(session, doc_id)
    if doc is None:
        raise PlatformError(ErrorCode.PDF_CORRUPT, f"document {doc_id} vanished")

    worker_id = f"parser-{uuid.uuid4().hex[:8]}"
    shard = repo.claim_shard(session, doc_id, idx, worker_id, s)
    session.flush()
    if shard is None:
        return  # someone else got it (§6.3 step 1)

    # Retry ladder (PRD §6.3) — attempts are not identical. claim_shard
    # already incremented attempts, so shard.attempts IS this attempt's number.
    if shard.attempts >= 2:
        ladder = _ladder_action(shard, s)
        if ladder == "split":
            return _split_and_requeue(session, redis, doc, shard, s)
        # "text_only": fall through with a degraded converter config

    tmpdir = Path(tempfile.mkdtemp(prefix="parse-"))
    pdf_path = tmpdir / "source.pdf"
    cache_dir = s.parser_pdf_cache_dir
    cache_path = Path(cache_dir) / f"{doc_id}.pdf" if cache_dir else None
    if cache_path is not None and cache_path.is_file():
        # Cache hit — PDFs are immutable per doc_id, no download needed.
        # Copy (not hardlink/ln): docling may write sidecar files next to the
        # source, and tmpfs unlinking on recycle must never touch the cache.
        shutil.copyfile(cache_path, pdf_path)
        logging.getLogger(__name__).info(
            "pdf cache HIT %s (%.1f MB saved download)",
            doc_id,
            cache_path.stat().st_size / 1e6,
        )
    else:
        s3.download_to(
            s3.make_s3(s), s.s3_bucket_raw, s3.raw_key(str(doc_id)), str(pdf_path)
        )
        if cache_path is not None:
            # Atomic write into the host cache: tmp file in the same dir,
            # then rename. Concurrent shard jobs of the same book may race —
            # os.replace is atomic, losers just overwrite with identical bytes.
            try:
                cache_path.parent.mkdir(parents=True, exist_ok=True)
                tmp_cache = cache_path.with_suffix(".tmp")
                shutil.copyfile(pdf_path, tmp_cache)
                os.replace(tmp_cache, cache_path)
                logging.getLogger(__name__).info(
                    "pdf cache MISS -> STORED %s", doc_id
                )
            except OSError as cache_exc:
                # Cache is a pure optimization — never fail the shard over it.
                logging.getLogger(__name__).warning(
                    "pdf cache store failed (ignored): %s", cache_exc
                )

    started = time.monotonic()
    verdict = needs_ocr(str(pdf_path), page_start, page_end, s)
    logging.getLogger(__name__).info(
        "ocr gate %s shard=%d pages=%d-%d mean_chars/page=%.0f needs_ocr=%s",
        doc_id,
        idx,
        page_start,
        page_end,
        verdict.mean_chars_per_page,
        verdict.needs_ocr,
    )
    # Ladder attempt >=4 (text_only): table structure off — a settings copy
    # so the shared s is untouched; converter_cache_key includes
    # parsing_do_table_structure, so the cache separates the variants for free.
    if shard.attempts >= 4 and s.parsing_do_table_structure:
        s = s.model_copy(update={"parsing_do_table_structure": False})
    cache_key = converter_cache_key(verdict.needs_ocr, s)
    cache_hit = cache_key in _CONVERTER_CACHE
    converter = get_converter(need_ocr=verdict.needs_ocr, settings=s, builder=build_converter)
    logging.getLogger(__name__).info(
        "converter cache %s need_ocr=%s key=%r",
        "HIT" if cache_hit else "MISS",
        verdict.needs_ocr,
        cache_key,
    )

    result = converter.convert(
        str(pdf_path),
        # docling 2.126: page_range is 1-based INCLUSIVE — (start, end) both
        # ≥1. Passing 0-based (page_start-1, ...) fails validation "start
        # must be ≥ 1" on every first shard. Overlap semantics unchanged
        # (page_end is inclusive both ways).
        page_range=(page_start, page_end),
    )
    check_rss_budget(s)
    markdown = result.document.export_to_markdown()

    parsed_key = s3.parsed_key(str(doc_id), idx)
    s3.upload_text(s3.make_s3(s), s.s3_bucket_parsed, parsed_key, markdown)

    duration_ms = int((time.monotonic() - started) * 1000)
    repo.mark_shard_done(
        session,
        doc_id,
        idx,
        duration_ms=duration_ms,
        peak_rss_mb=check_rss_budget(s),
        parsed_uri=f"s3://{s.s3_bucket_parsed}/{parsed_key}",
        needs_ocr=verdict.needs_ocr,
    )

    # last shard settled → enqueue embed (§6.3 step 7).
    # mark_shard_done() increments shards_done via SQL expression; refresh
    # the ORM row before checking, otherwise the final shard appears missing.
    session.flush()
    session.expire(doc)
    doc = repo.get_document(session, doc_id)
    if repo.book_settled(doc):
        streams.xadd_job(redis, streams.STREAM_EMBED, contracts.EmbedJob(doc_id=doc_id))


def main() -> None:  # pragma: no cover - process entry
    import logging

    from core.db.session import make_engine, make_session_factory
    from core.observability.logging import configure_logging

    settings = get_settings()
    configure_logging(settings.log_level)
    logging.getLogger(__name__).info("parser worker starting")

    factory = make_session_factory(make_engine(settings))
    redis = streams.make_redis(settings)

    def handler(session, job):
        handle_parse(session, job, redis, settings)

    from workers.runner import record_job_error, run_consumer

    run_consumer(
        stream=streams.STREAM_PARSE,
        consumer=streams.new_consumer_name("parser"),
        handler=handler,
        session_factory=factory,
        redis=redis,
        prefetch=1,  # §6.3: one job per process, non-negotiable
        on_error=lambda s, j, e: record_job_error(s, j, e, stage="parse"),
        recycle_after=settings.parser_recycle_after,
    )


if __name__ == "__main__":
    main()

File: internal/service/parser.go
package service

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"time"

	"github.com/jackc/pgx/v5"

	"github.com/saufi-opi/lybrix/internal/errors"
	"github.com/saufi-opi/lybrix/internal/objectstore"
	"github.com/saufi-opi/lybrix/internal/pipeline"
	"github.com/saufi-opi/lybrix/internal/queue"
	"github.com/saufi-opi/lybrix/internal/store"
)

// ladderAction decides the retry ladder step (parser.py _ladder_action):
// attempt 2: quarter-split; attempt 3: single-page; attempt >=4: text-only.
// A shard too small to split skips straight to text-only.
// Returns "split"|"text_only".
//
// EPUB rule (PLAN.md item 19): a span-1 shard (every EPUB synthetic shard)
// already fails both split conditions and falls through to text_only —
// retries re-attempt the same single shard with docling backoff, and no
// duplicate sub-shards are ever created. The explicit span-1 guard below
// pins that contract against future threshold drift.
func LadderAction(shard *store.Shard, shardPages int) string {
	span := shard.PageEnd - shard.PageStart + 1
	if span < 2 {
		return "text_only" // single-page/EPUB synthetic shard: never split
	}
	if shard.Attempts == 2 && span >= 4*(maxInt(1, shardPages/4))/2 { // worth quarter-splitting
		return "split"
	}
	if shard.Attempts == 3 && span >= 2 {
		return "split" // single-page sub-shards
	}
	return "text_only"
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// shouldSlicePDF decides whether HandleParse cuts the shard's page range out
// of the source PDF before parsing (BACKLOG R-22). Guards, in order:
//
//   - only real PDF shards slice — non-PDF shards are synthetic single
//     shards, the whole file is their unit;
//   - both range bounds must be known (0 means "no range in the job") and
//     well-ordered (pageEnd >= pageStart, matching ParseJob.ValidatePages'
//     invariant);
//   - a shard spanning the entire document is passed through untouched
//     (no pointless copy).
func shouldSlicePDF(docFormat pipeline.Format, pageStart, pageEnd, totalPages int) bool {
	if docFormat != pipeline.FmtPDF {
		return false
	}
	if pageStart <= 0 || pageEnd < pageStart {
		return false
	}
	return !(pageStart == 1 && pageEnd >= totalPages)
}

// SplitAndRequeue replaces a failing shard with finer sub-shards (ladder
// attempts 2-3). The parent becomes SKIPPED; each sub-shard is a fresh
// ParseJob. Port of parser.py _split_and_requeue.
func SplitAndRequeue(ctx context.Context, deps Deps, tx pgx.Tx, doc *store.Document, shard *store.Shard) error {
	span := shard.PageEnd - shard.PageStart + 1
	shardPages := 1
	if shard.Attempts < 3 {
		shardPages = maxInt(1, span/4)
	}
	fb, err := pipeline.FixedBounds(span, shardPages, 0)
	if err != nil {
		return err
	}
	// overlap=0 is required: sub-shards must tile the parent disjointly
	// (fixed_bounds' default overlap=1 would emit overlapping bounds —
	// a span-20 attempt-2 shard would yield 5 overlapping bounds like
	// [1-5],[5-9],[9-13],... instead of 4 disjoint ones, double-counting
	// total_shards and re-parsing boundary pages).
	bounds := make([][2]int, 0, len(fb))
	for _, b := range fb {
		bounds = append(bounds, [2]int{shard.PageStart + b.PageStart - 1, shard.PageStart + b.PageEnd - 1})
	}
	startIdx, err := deps.DB.NextShardIdx(ctx, tx, doc.ID)
	if err != nil {
		return err
	}
	if err := deps.DB.SkipShard(ctx, tx, doc.ID, shard.Idx); err != nil {
		return err
	}
	if _, err := deps.DB.InsertShards(ctx, tx, doc.ID, bounds, startIdx); err != nil {
		return err
	}
	newTotal := (derefInt(doc.TotalShards)) + len(bounds)
	if _, err := tx.Exec(ctx,
		`UPDATE documents SET total_shards = $2, updated_at = NOW() WHERE id = $1`, doc.ID, newTotal); err != nil {
		return err
	}
	for i, b := range bounds {
		job := queue.ParseJob{
			SchemaVersion: queue.SchemaVersion,
			DocID:         doc.ID,
			Idx:           startIdx + i,
			PageStart:     b[0],
			PageEnd:       b[1],
			SourceURI:     doc.SourceURI,
		}
		if _, err := queue.XAddJob(ctx, deps.Redis, queue.StreamParse, job); err != nil {
			return err
		}
	}
	return store.WriteEvent(ctx, tx, "warn", "parse",
		fmt.Sprintf("shard %d re-split into %d sub-shards (ladder attempt %d)", shard.Idx, len(bounds), shard.Attempts),
		strPtr(doc.ID), intPtr(shard.Idx), strPtr("SHARD_RESPLIT"), nil, nil)
}

// HandleParse is the parser handler: THE heavy one (PRD §6.3).
//
// Memory discipline, all five controls:
//
//	one job per process (prefetch=1)        — compose env + runner wiring
//	process recycling (PARSER_RECYCLE_AFTER) — compose + runner
//	tmpfs cap                                — compose
//	retry ladder                             — below
//	soft RSS (no-op in Go; recording stays)  — pipeline.CurrentRSSMB
func HandleParse(ctx context.Context, deps Deps, tx pgx.Tx, job map[string]any) error {
	s := deps.Settings
	docID, err := jobString(job, "doc_id")
	if err != nil {
		return err
	}
	idx, ok := jobInt(job, "idx")
	pageStart, _ := jobInt(job, "page_start")
	pageEnd, _ := jobInt(job, "page_end")
	if !ok {
		return fmt.Errorf("job field %q missing", "idx")
	}

	doc, err := deps.DB.GetDocument(ctx, docID)
	if err != nil {
		return err
	}
	if doc == nil {
		return errors.NewPlatformError(errors.CodePDFCorrupt, fmt.Sprintf("document %s vanished", docID))
	}

	workerID := fmt.Sprintf("parser-%s", shortID())
	shard, err := deps.DB.ClaimShard(ctx, tx, docID, idx, workerID, s.ShardLeaseSeconds)
	if err != nil {
		return err
	}
	if shard == nil {
		return nil // someone else got it (§6.3 step 1)
	}

	// Retry ladder (PRD §6.3) — attempts are not identical. claim_shard
	// already incremented attempts, so shard.attempts IS this attempt's
	// number.
	textOnly := false
	if shard.Attempts >= 2 {
		action := LadderAction(shard, s.ShardPages)
		if action == "split" {
			return SplitAndRequeue(ctx, deps, tx, doc, shard)
		}
		// "text_only": fall through with a degraded converter config
		textOnly = shard.Attempts >= 4
	}

	tmpDir, err := os.MkdirTemp("", "parse-")
	if err != nil {
		return err
	}
	defer os.RemoveAll(tmpDir)
	// DocFormat from the doc's mime type — the shared format table is the
	// single source of truth (Workstream 2). Unknown mime → PDF behavior.
	docFormat := pipeline.FormatFromMime(derefStr(doc.MimeType))
	sourceName := pipeline.ParseSourceName(docFormat)
	sourcePath := filepath.Join(tmpDir, sourceName)
	// PDF cache path: bypassed for non-PDF docs (office/text docs are
	// ≤64 MiB — always download; the cache dir holds .pdf files and the
	// cache path derivation is doc-scoped, not mime-scoped).
	cachePath := ""
	if s.ParserPDFCacheDir != "" && docFormat == pipeline.FmtPDF {
		cachePath = pipeline.PDFCachePath(s.ParserPDFCacheDir, docID)
	}
	if cachePath != "" {
		if _, err := os.Stat(cachePath); err == nil {
			// Cache hit — PDFs are immutable per doc_id, no download needed.
			// Copy (not hardlink/ln): tmpfs unlinking on recycle must never
			// touch the cache.
			if err := copyFileLocal(sourcePath, cachePath); err != nil {
				return err
			}
			slog.Info("pdf cache HIT", "doc", docID)
		} else {
			if err := deps.S3.DownloadTo(ctx, s.S3BucketRaw, objectstore.RawKey(docID), sourcePath); err != nil {
				return errors.NewPlatformError(errors.CodePDFCorrupt, fmt.Sprintf("source missing: %v", err))
			}
			pipeline.StorePDFCache(s.ParserPDFCacheDir, docID, sourcePath)
			slog.Info("pdf cache MISS -> STORED", "doc", docID)
		}
	} else {
		if err := deps.S3.DownloadTo(ctx, s.S3BucketRaw, objectstore.RawKey(docID), sourcePath); err != nil {
			return errors.NewPlatformError(errors.CodePDFCorrupt, fmt.Sprintf("source missing: %v", err))
		}
	}

	// Shard slicing (BACKLOG R-22): ParseRequest's page range is consumed by
	// NO parser — docling and anydoc both convert whatever file they are
	// handed end to end. For PDF shards that are not the whole document, cut
	// pageStart..pageEnd out with pdfcpu first so each shard converts only
	// its own pages. Non-PDF shards are synthetic single shards; the whole
	// file is the unit (types.go IsSynthetic).
	parsePath := sourcePath
	reqStart, reqEnd := pageStart, pageEnd
	if docFormat == pipeline.FmtPDF && pageStart > 0 && pageEnd >= pageStart {
		totalPages, perr := pipeline.PageCount(sourcePath)
		if perr != nil {
			// PageCount failure = the PDF may be corrupt/encrypted; leave the
			// request untouched and let the existing conversion/error paths
			// classify it.
			slog.Debug("page count failed; skipping slice", "err", perr, "doc", docID)
		} else if shouldSlicePDF(docFormat, pageStart, pageEnd, totalPages) {
			sliced := filepath.Join(tmpDir, fmt.Sprintf("shard-%d.pdf", idx))
			n, serr := pipeline.SlicePDF(sourcePath, sliced, pageStart, pageEnd)
			if serr == nil && n > 0 {
				parsePath = sliced
				// Renormalize: the slice IS pages 1..n now. The OCR gate inside
				// TwoTierParser probes req.PDFPath at req.PageStart..req.PageEnd,
				// so it must probe the slice.
				reqStart, reqEnd = 1, n
				slog.Info("pdf shard sliced",
					"doc", docID, "shard", idx,
					"pages", fmt.Sprintf("%d-%d/%d", pageStart, pageEnd, totalPages))
			} else if serr != nil {
				// Slice failure must not fail the shard — degrade to the 1.0
				// behavior (whole-file conversion) and let the ladder handle a
				// repeat failure.
				slog.Warn("pdf slice failed; parsing whole file", "err", serr,
					"doc", docID, "shard", idx)
			}
		}
	}

	started := time.Now()
	result, err := deps.Parser.Parse(ctx, pipeline.ParseRequest{
		PDFPath:    parsePath,
		PageStart:  reqStart,
		PageEnd:    reqEnd,
		Attempt:    shard.Attempts,
		ShardPages: s.ShardPages,
		TextOnly:   textOnly,
		SkipAnyDoc: docFormat == pipeline.FmtEPUB,
		DocFormat:  docFormat,
	})
	if err != nil {
		return err
	}
	slog.Info("parse done",
		"doc", docID, "shard", idx, "pages", fmt.Sprintf("%d-%d", pageStart, pageEnd),
		"engine", result.Engine, "mean_chars/page", result.MeanCharsPerPage,
		"needs_ocr", result.NeedsOCR, "duration_ms", result.DurationMS)

	parsedKey := objectstore.ParsedKey(docID, idx)
	if err := deps.S3.UploadText(ctx, s.S3BucketParsed, parsedKey, result.Markdown, "text/markdown"); err != nil {
		return err
	}
	durationMS := int(time.Since(started).Milliseconds())
	if result.DurationMS > 0 {
		durationMS = int(result.DurationMS)
	}
	if err := deps.DB.MarkShardDone(ctx, tx, docID, idx, durationMS, result.PeakRSSMB,
		"s3://"+s.S3BucketParsed+"/"+parsedKey, result.NeedsOCR, floatPtr(result.MeanCharsPerPage)); err != nil {
		return err
	}

	// last shard settled → enqueue embed (§6.3 step 7). MarkShardDone
	// increments shards_done via SQL; re-read the doc row through the same tx
	// before checking — a pool read cannot see the uncommitted increment, so
	// the final shard would appear missing and embedding would wait for the
	// janitor's next pass.
	fresh, err := deps.DB.GetDocumentTx(ctx, tx, docID)
	if err != nil {
		return err
	}
	if fresh != nil && store.BookSettled(fresh) {
		_, err := queue.XAddJob(ctx, deps.Redis, queue.StreamEmbed, queue.EmbedJob{
			SchemaVersion: queue.SchemaVersion, DocID: docID,
		})
		return err
	}
	return nil
}

func floatPtr(f float64) *float64 { return &f }

func copyFileLocal(dst, src string) error {
	b, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, b, 0o644)
}
